### PR TITLE
F #3930: NSX Security Groups

### PR DESCRIPTION
**Pull request by mouyaq**
Security groups working on NSX-T and NSX-V
Pending: Subprotocols ICMP ( pending to add, right now all ICMP is added to rules )
Pending: Changes to improve error detections
Pending: Testing

### CODE REVIEW

File: src/vmm_mad/remotes/lib/nsx_driver/distributed_firewall.rb [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+
+        include NSXDriver::NSXRule
+
+        # ATTRIBUTES
+        attr_reader :one_section_name
+
+        # CONSTRUCTOR
+        def initialize(nsx_client)
+            super(nsx_client)
+            @one_section_name = NSXConstants::ONE_SECTION_NAME
```
[the remark this answers is not in the document]
Comment (mouyaq) [reply src/vmm_mad/remotes/lib/nsx_driver/distributed_firewall.rb]: @one_section_name is not used right now, i'm going to delete it.

File: src/vmm_mad/remotes/lib/nsx_driver/distributed_firewall.rb [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+                    suffix = device.key.to_s[1..-1]
+                    lpid = "#{vm.item.config.instanceUuid}.#{suffix}"
+                    nic_lp = lpid
+                else
+                    error_msg = "Network type is: #{network_pgtype} \
+                                    and should be \
+                                    #{NSXConstants::NSXT_LS_TYPE} \
+                                    or #{NSXConstants::NSXV_LS_TYPE}"
+                    error = NSXError::UnknownObject.new(error_msg)
+                    STDERR.puts error_msg
```
[the remark this answers is not in the document]
Comment (mouyaq) [reply src/vmm_mad/remotes/lib/nsx_driver/distributed_firewall.rb]: remove STDERR the error is raised and should be caught at post script

File: src/vmm_mad/remotes/lib/nsx_driver/distributed_firewall.rb [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+                    sg_rules_array.each do |sg_spec|
+                        create_rule(sg_spec)
+                    end
+                end
+            end
+        end
+
+        def clear_all_rules(template)
+            template_xml = Nokogiri::XML(template)
+            vm_id = template_xml.xpath('//VM/ID').text
```
[the remark this answers is not in the document]
Comment (mouyaq) [reply src/vmm_mad/remotes/lib/nsx_driver/distributed_firewall.rb]: Changed to /VM/ID

File: src/vmm_mad/remotes/lib/nsx_driver/distributed_firewall.rb [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+            vm_deploy_id = template_xml.xpath('//DEPLOY_ID').text
```
[the remark this answers is not in the document]
Comment (mouyaq) [reply src/vmm_mad/remotes/lib/nsx_driver/distributed_firewall.rb]: Changed to /VM/ID

[the remark this answers is not in the document]
Comment (mouyaq) [reply src/vmm_mad/remotes/lib/nsx_driver/distributed_firewall.rb]: Changed to /VM/DEPLOY_ID

File: src/vmm_mad/remotes/lib/nsx_driver/distributed_firewall.rb [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+                # network_id = nic.xpath('NETWORK_ID').text
+                sec_groups = nic.xpath('SECURITY_GROUPS').text.split(',')
+                sec_groups.each do |sec_group|
+                    # Get Security Group ID and NAME
+                    xp = "//SECURITY_GROUP_RULE[SECURITY_GROUP_ID=#{sec_group}]"
+                    sg_id = template_xml.xpath(xp)[0]
+                                        .xpath('SECURITY_GROUP_ID').text
+                    sg_name = template_xml.xpath(xp)[0]
+                                          .xpath('SECURITY_GROUP_NAME').text
+                    rule_name =  "#{sg_id}-#{sg_name}-#{vm_id}"
```
[the remark this answers is not in the document]
Comment (mouyaq) [reply src/vmm_mad/remotes/lib/nsx_driver/distributed_firewall.rb]: # Class DistributedFirewall
    # Naming convention for rules names:
    # "<sg_id>-<sg_name>-<vm_id>-<vm_deploy_id>-<nic_id>"
    # <sg_id> : Security Group ID
    # <sg_name> : Security Group name
    # <vm_id> : Virtual Machine ID
    # <vm_deploy_id> : Vmware virtual machine reference ( vm-XXX )
    # <nic_id> : NIC ID in the virtual machine

File: src/vmm_mad/remotes/lib/nsx_driver/nsxv_dfw.rb [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+            url = @url_sections + '/' + section_id + '/rules/' + rule_id
+            # etag is needed to add a new header If-Match
+            etag = section_etag(section_id)
+            raise "Cannot get etag from section: #{section_id}" unless etag
+
+            aditional_headers = [{ 'If-Match' => etag }]
+            @nsx_client.delete(url, aditional_headers)
+            # Aditional check
+            # result = rule_by_id(rule_id)
+            # raise 'Error deleting rule in DFW' unless result.empty?
```
[the remark this answers is not in the document]
Comment (mouyaq) [reply src/vmm_mad/remotes/lib/nsx_driver/nsxv_dfw.rb]: removed

File: src/vmm_mad/remotes/lib/nsx_driver/nsxv_logical_port.rb [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+#                                                                            #
+# Unless required by applicable law or agreed to in writing, software        #
+# distributed under the License is distributed on an "AS IS" BASIS,          #
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   #
+# See the License for the specific language governing permissions and        #
+# limitations under the License.                                             #
+#--------------------------------------------------------------------------- #
+module NSXDriver
+
+    # Class NSXVLogicalPort
```
[the remark this answers is not in the document]
Comment (mouyaq) [reply src/vmm_mad/remotes/lib/nsx_driver/nsxv_logical_port.rb]: rubocop complains if I remove the comment

File: src/vmm_mad/remotes/lib/nsx_driver/nsxv_client.rb [hunk 8]
```diff
@@ -47,130 +47,122 @@ class NSXVClient < NSXDriver::NSXClient
...
-            rescue StandardError => e
-                raise e
-            end
-            return response \
-                if check_response(response, [NSXDriver::NSXConstants::CODE_OK])
+            response = Net::HTTP.start(uri.host, uri.port, :use_ssl => true,
+                :verify_mode => OpenSSL::SSL::VERIFY_NONE) do |https|
+                    https.request(request)
+                end
+            response = check_response(response, valid_codes)
```
[the remark this answers is not in the document]
Comment (mouyaq) [reply src/vmm_mad/remotes/lib/nsx_driver/nsxv_client.rb]: I think leave check_response as last line can be confused, maybe it's better change the function name

[the remark this answers is not in the document]
Comment (mouyaq) [reply src/vmm_mad/remotes/lib/nsx_driver/nsxv_client.rb]: removed

File: src/vmm_mad/remotes/lib/nsx_driver/nsxv_client.rb [hunk 9]
```diff
@@ -47,130 +47,122 @@ class NSXVClient < NSXDriver::NSXClient
...
             request.basic_auth(@nsx_user, @nsx_password)
             response = Net::HTTP.start(uri.host, uri.port, :use_ssl => true,
-              :verify_mode => OpenSSL::SSL::VERIFY_NONE) do |https|
-                  https.request(request)
-              end
-            check_response(response, [NSXDriver::NSXConstants::CODE_OK])
+                :verify_mode => OpenSSL::SSL::VERIFY_NONE) do |https|
+                    https.request(request)
+                end
+            response = check_response(response, valid_codes)
```
[the remark this answers is not in the document]
Comment (mouyaq) [reply src/vmm_mad/remotes/lib/nsx_driver/nsxv_client.rb]: removed var

File: src/vmm_mad/remotes/lib/nsx_driver/nsxv_dfw.rb [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+        # - rule | nil
+        def rule_by_id(rule_id, section_id = @one_section_id)
+            url = @url_sections + '/' + section_id + '/rules/' + rule_id
+            valid_codes = [NSXConstants::CODE_CREATED,
+                           NSXConstants::CODE_OK,
+                           NSXConstants::CODE_BAD_REQUEST,
+                           NSXConstants::CODE_NOT_FOUND]
+            additional_headers = []
+            result = @nsx_client.get(url, additional_headers, valid_codes)
+            rule = result.xpath(NSXConstants::NSXV_DFW_RULE_XPATH)
```
[the remark this answers is not in the document]
Comment (mouyaq) [reply src/vmm_mad/remotes/lib/nsx_driver/nsxv_dfw.rb]: removed var

File: src/vmm_mad/remotes/lib/nsx_driver/nsxv_logical_port.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (tinova) [reply src/vmm_mad/remotes/lib/nsx_driver/nsxv_logical_port.rb]: this should be a general description of the class

# This class implements the representation of a NSX Logical Port

adding any particular details that ease the readibility of the code

**Pull request merged**